src/transform/src/typecheck.rs

```diff
@@ -27,11 +26,11 @@ use mz_repr::{ColumnName, ColumnType, RelationType, Row, ScalarBaseType, ScalarT
 ///
 /// We use a `RefCell` to ensure that contexts are shared by multiple typechecker passes.
 /// Shared contexts help catch consistency issues.
-pub type SharedContext = Rc<RefCell<Context>>;
```

Comment:
Is it possible to avoid this turning into an `Arc<Mutex<_>>`? Costs of atomics and mutexes are much higher than for `Rc<RefCell<_>>` and it seems we don't actually want multiple threads manipulating this in parallel?

Reply:
> it seems we don't actually want multiple threads manipulating this in parallel

No, but we want to be able to pass an `Optimize` implementation through a thread boundary in order to do off-thread optimization, so we need all its fields to be `Send + Sync`.

If you look for the remaining changes in the same commit you will see that there is only a couple of places where we actually do `.lock().unwrap()` on that value and they are not happening frequently on the hot path (it's done more when you initialize the transform that needs the `Context`).

Reply:
Here is a scenario that I can think of which does not do multi-threaded optimization but demonstrates why the optimizer might need to be shared shared across thread boundaries:

- The main thread creates an optimizer.
- It spawns a thread that calls the first two `optimizer.optimize()` stages with `tokio::task::spawn`.
- It joins the future, obtaining the result of the second stage, and does a thread-local call to do some timestamp resolution.
- It spawns a thread that calls the next two `optimizer.optimize()` stages with `tokio::task::spawn`.

I think this is more or less how the optimization path for the `optimize_peek` will look once I add it. I think once we introduce the new distributed timestamp oracle this will no longer apply, at which point we can revisit this decision.

Reply:
Here is a scenario that I can think of which does not do multi-threaded optimization but demonstrates why the optimizer might need to be shared shared across thread boundaries:

- The main thread creates an optimizer.
- It spawns a thread that calls the first two `optimizer.optimize()` stages with `tokio::task::spawn`.
- It joins the future, obtaining the result of the second stage, and does a thread-local call to do some timestamp resolution.
- It spawns a thread that calls the next two `optimizer.optimize()` stages with `tokio::task::spawn`.

I think this is more or less how the optimization path for the `optimize_peek` will look once I add it. I think once we introduce the new distributed timestamp oracle this will no longer apply, at which point we can revisit this decision.